File: leetcode/easy/217-Contains-Duplicate.py
#!/usr/bin/python3

class Solution:
  # Solution 1: Using dictionaries
  def containsDuplicate(self, nums: list) -> bool:
    # Initialize a dictionary
    m = {}

    # Iterate through the values in the given list. If 
    # its already in the dictionary then we know theres
    # a duplicate. Otherwise we continue through the 
    # entire list.
    for n in nums:
      if n in m:
        return True
      m[n] = n
    return False

  # Solution 2: Using sets
  def containsDuplicate(self, nums: list) -> bool:
    s = set(nums)
    return not len(s) == len(nums)

File: neetcode150/easy/cpp/217-Contains-Duplicate.cpp
#include <vector>
#include <set>
using namespace std;

// Solution 1
// Note: Slow in C++
class Solution{
public:
  // Check if the length of the set 
  bool containsDuplicate(vector<int>& nums){
    set<int> s;
    for(int n : nums){
      s.insert(n);
    }
    return s.size() != nums.size();
  }
};

// Solution 2
class Solution{
public:
  bool containsDuplicate(vector<int>& nums){
    // Sort the vector
    sort(nums.begin(), nums.end());
    int x = nums[0];
    // if n == n + 1 then the vector contains a duplicate
    for(int i = 1; i < nums.size(); i++){
      if(x == nums[i]){
        return true;
      }
      // set n = n + 1
      x = nums[i];
    }
    return false;
  }
};